distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso

```diff
@@ -261,54 +263,65 @@ type Vector a
                         @Tail_Call fold_function (function current (self.at idx)) (idx + 1)
                 fold_function (self.at 0) 1
 
-    ## Computes the sum of the values in the vector.
+    ## Returns the first element of the vector that satisfies the predicate or
+       if no elements of the vector satisfy the predicate, it throws nothing.
 
-       For this method to be defined, the elements of the vector must be able to
-       have the `+` operator used to combine them.
+       Arguments:
+       - predicate: A function that takes a list element and returns a boolean
+         value that says whether that value satisfies the conditions of the
+         function.
+       - start: The index to start searching from.
+       - if_missing: Value returned if no element satisfies the predicate.
 
        > Example
-         Compute the sum of all fo the elements in a vector.
+         Finding a first element of the vector that is larger than 3.
 
-             [0, 1, 2].sum
-    sum : Any ! (Empty_Error | No_Such_Method)
-    sum self =
-        result = Panic.recover Any <| self.reduce (+)
-        result.map_error x->case x of
-            _ : No_Such_Method -> x
-            Empty_Error -> x
-            _ -> Panic.throw x
+             [1, 2, 3, 4, 5].find (> 3)
+    find : (Any -> Boolean) -> Integer -> Any -> Any
+    find self predicate start=0 ~if_missing=(Error.throw Not_Found) =
+        used_start = 0.max (if start < 0 then start + self.length else start)
+        found = used_start.up_to self.length . find (idx -> (predicate (self.at idx)))
+        if found.is_nothing then if_missing else self.at found
 
-    ## Checks whether a predicate holds for at least one element of this vector.
+    ## Returns the index of an element in the vector. Returns -1 if the element
+       is not found.
 
        Arguments:
-       - predicate: A function that takes a list element and returns a boolean
-         value that says whether that value satisfies the conditions of the
-         function.
+       - element: The element to search for or a predicate function to test for
+         each element.
+       - start: The index to start searching from.
 
        > Example
-         Check if any element of the list is larger than 3.
+         Find the index of an element in a vector.
 
-             [1, 2, 3, 4, 5].exists (> 3)
-    exists : (Any -> Boolean) -> Boolean
-    exists self predicate =
-        0.up_to self.length . exists (idx -> (predicate (self.at idx)))
+             [1, 2, 3].index_of 2 == 1
+    index_of : (Any -> Boolean) -> Integer -> Any
+    index_of self element start=0 =
+        predicate = case element of
+            _ : Function -> (i -> element (self.at i))
+            _ -> (i -> self.at i == element)
+        found = start.up_to self.length . find predicate
+        found.if_nothing -1
```

Comment:
```suggestion
        found
```

What about just returning `Nothing`?

I know it is quite a common convention to return a `-1` but IMHO it isn't very great as it uses a magic value to represent a lack of result. Seems to me that returning (_not throwing_) a `Nothing` would represent the fact that the element has not been found would be much more meaningful.

Especially because in most of our APIs `-1` _is_ a **valid** index. In most languages that return `-1` to indicate that the element is missing, the `-1` is a completely invalid index so it can be used as such indicator (still IMO if possible a special value like Nothing is better). But in Enso `-1` is actually valid, so it is even more confusing to return it meaning that the element has not been found. I could interpret it as 'element was found on last position'. 

Ofc documentation is explaining this well, but still if you think it's OK I'd suggest switching to Nothing which is more meaningful and less error-prone, at least IMO.

Reply:
agree - will return Nothing.